src/vm/virtualMachine.ts

```diff
@@ -351,4 +369,58 @@ export class VirtualMachine {
         break;
     }
   }
+
+  /** Relocate memory and trace */
+  relocate() {
+    let relocationTable = [1];
+    let memorySize = 1;
+
+    for (const segment of this.memory.values) {
+      memorySize += segment.length;
+      relocationTable.push(memorySize);
+    }
```

Comment:
so your `relocatationTable` is the start index of the segment in the final flatten memory

hence you want to do instead of cumsum of the segment lengths

```
this.memory.values
    .map((segment) => segment.length)
    .map((sum => value => sum += value)(0))
```

Reply:
Yes, so `relocationTable` must starts with the offset of the segment 0, which is 1 (the relocated memory is 1-based...)

Mapping each segment to its length and directly do the cumulative sum will not include the offset of segment 0 and the correct offset for a 1-based memory

To use `.map` and have the correct relocation table we can do this:
```typescript
const relocationTable = this.memory.values
  .map((segment) => segment.length)
  .map(
    ((sum) => (value, index) =>
      index ? (sum += value) : [1, (sum += value)])(1)
  )
  .flat();
```

Or this, but I prefer the first version:

```typescript
const relocationTable = [
  1,
  ...this.memory.values
    .map((segment) => segment.length)
    .map(
      (
        (sum) => (value) =>
          (sum += value)
      )(1)
    ),
];
```

Reply:
Actually the offset of segment 0 was included but miscalculated (adding its length to its final base address when it shouldn't)

A working solution would be:
```typescript
const relocationTable = this.memory.values
  .map((segment) => segment.length)
  .map(
    (
      (sum) => (value) =>
        (sum += value) - value
    )(1)
  );
```